<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [414]:
import pandas as pd
import psycopg2


In [415]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [416]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [417]:
# код запроса представляет из себя строковую переменную
n = 10 
query = f'''select * 
            from vacancies
                 
            limit {n}
         '''

In [418]:
df = pd.read_sql_query(query, connection) # используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\3918204492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection) # используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма


id                                               name  \
0  55312386                                Компьютерный Мастер   
1  55843014                            Системный администратор   
2  54525964                      Lead Java Developer to Poland   
3  54525965                      Lead Java Developer to Poland   
4  55354053  Специалист службы поддержки с техническими зна...   
5  55906072                                          Корректор   
6  55523207                   React Front-end Middle Developer   
7  55050261                                Инженер-программист   
8  55583432                            Системный администратор   
9  55312414        Ведущий специалист группы разработки систем   

                                          key_skills          schedule  \
0  Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1  Средства криптографической защиты информации\t...       Полный день   
2    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                               None  Удаленная работа   
5  Корректура текстов\tГрамотность\tРабота в кома...       Полный день   
6       JavaScript\tTypeScript\tCSS3\tNode.js\tHTML5     Гибкий график   
7  Java SE\tSpring Framework\tSpring Boot\tGit\tS...       Полный день   
8  Linux\tАдминистрирование сетевого оборудования...       Полный день   
9  1С программирование\t1С: Предприятие 8\t1С: До...       Полный день   

           experience           employment  salary_from  salary_to  area_id  \
0           Нет опыта     Полная занятость      64000.0        NaN     1516   
1       От 3 до 6 лет     Полная занятость          NaN        NaN      130   
2       От 3 до 6 лет     Полная занятость          NaN        NaN      160   
3       От 3 до 6 лет     Полная занятость          NaN        NaN      159   
4           Нет опыта  Частичная занятость      15000.0        NaN     1955   
5           Нет опыта     Полная занятость          NaN        NaN     2323   
6  От 1 года до 3 лет     Полная занятость          NaN        NaN     2759   
7           Нет опыта     Полная занятость      40000.0    60000.0       49   
8  От 1 года до 3 лет     Полная занятость      50000.0        NaN        3   
9       От 3 до 6 лет     Полная занятость          NaN        NaN       47   

   employer_id  
0      5724811  
1      4903713  
2        69961  
3        69961  
4         1740  
5       584934  
6       204511  
7      4448636  
8        55126  
9      1413754

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [419]:

query_3_1 = f'''select count(*) as count_vacancies
                from vacancies
                 
         
           '''

In [420]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count_vacancies
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [421]:

query_3_2 = f'''select count(*) as count_employers
                from employers 
                 
         
           '''

In [422]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count_employers
0            23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [423]:

query_3_3 = f'''select count(*) as count_areas
                from areas
                 
         
           '''

In [424]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count_areas
0         1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [425]:
query_3_4 = f'''select count(*) as count_industries
                from industries
                 
         
             '''

In [426]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count_industries
0               294

***

В базе содержится 49197 вакансий по 294 сферам деятельности. Данные вакансии представлены 23501 работодателем, которые находятся 1362 регионах. 





# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [427]:
 
query_4_1 = f'''select 
                    a.name as area, 
                    count(*) as cnt_vacancies
                from areas a
                join vacancies v on a.id  = v.area_id 
                group by 1  
                order by 2 desc 
                limit 10
            '''


In [428]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area  cnt_vacancies
0           Москва           5333
1  Санкт-Петербург           2851
2            Минск           2112
3      Новосибирск           2006
4           Алматы           1892
5     Екатеринбург           1698
6  Нижний Новгород           1670
7           Казань           1415
8        Краснодар           1301
9           Самара           1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [429]:
# текст запроса
query_4_2 = f'''select count(*) as count_vacancies
                from vacancies 
                where salary_from IS NOT NULL OR 
                salary_to IS NOT NULL
            '''

In [430]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count_vacancies
0            24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [431]:

query_4_3 = f'''select 
                    round(avg(salary_from)) as avg_salary_from,
                    round(avg(salary_to)) as avg_salary_to
                from vacancies
            '''

In [432]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [433]:

query_4_4 = f'''select 
                    schedule as count_schedule,
                    employment as count_employment,
                    count(*) as count_vacancies
                from vacancies
                group by 1, 2
                order by 3 desc
            '''

In [434]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


count_schedule     count_employment  count_vacancies
0        Полный день     Полная занятость            35367
1   Удаленная работа     Полная занятость             7802
2      Гибкий график     Полная занятость             1593
3   Удаленная работа  Частичная занятость             1312
4     Сменный график     Полная занятость              940
5        Полный день           Стажировка              569
6     Вахтовый метод     Полная занятость              367
7        Полный день  Частичная занятость              347
8      Гибкий график  Частичная занятость              312
9        Полный день     Проектная работа              141
10  Удаленная работа     Проектная работа              133
11     Гибкий график           Стажировка              116
12    Сменный график  Частичная занятость              101
13  Удаленная работа           Стажировка               64
14     Гибкий график     Проектная работа               18
15    Сменный график           Стажировка               12
16    Вахтовый метод     Проектная работа                2
17    Сменный график     Проектная работа                1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [435]:
query_4_5 = f'''select 
                    experience, 
                    count(id) as count_vacancies
                from vacancies
                group by experience
                order by 2 ASC
            '''

In [436]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count_vacancies
0         Более 6 лет             1337
1           Нет опыта             7197
2       От 3 до 6 лет            14511
3  От 1 года до 3 лет            26152

***

- Наибольшее количество вакансий предлагается в г.Москва - 5333, или 10,8% от общего количества вакансий, в пятерку лидеров также входят 
Санкт - Петербург, Минск, Новосибриск и Алматы. Количество вакансий зависит от размеров и численности населенных пунктов, в которых они представлены. 
- Часть работодателей не заполняют поле "заработная плата". Соответствующее поле заполнено только в 24073 вакансиях, что составляет - 49% от общего количетсва вакансий. К причинам отсутсвия заполненого поля "заработная плата" можно отнести неопределенность работодателя с уровнем заработной платы, трудности с определением объема работы, зависимость заработной платы от уровня кандидата, работодатель ищет топ специалистов и не разглашает уровень заработной платы. 
- Средняя заработная плата варьируется от 71065 руб. до 110537 руб. Разница между нижней и верхней границей зарплатной вилки составляет 39472руб. 
- Для работодателей в приоритете тип трудоустройста и рабочего графика является полная занятость и полный рабочий день, вакансии с такими типами составляют 72% от общего количества вакансий, однако сочетание удаленной работы и полного графика также является востребованным, так как составляет 16% от общего числа вакансий. 
- Работодатели чаще всего отдают предпочтение опытным кандитам, так как наибольшее количество вакансий предлагается для категории соискателей, имеющих опыт работы от 1 года до 3 лет, а также от 3 до 6 лет. 


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [437]:
query_5_1 = f'''select e.name as name_employers, 
                       count(v.id) as count_vacancies
                from employers e
                join vacancies v on e.id = v.employer_id
                group by 1
                order by 2 desc
                limit 5
            '''

In [438]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name_employers  count_vacancies
0         Яндекс             1933
1     Ростелеком              491
2       Тинькофф              444
3           СБЕР              428
4  Газпром нефть              331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [439]:
query_5_2 = f'''select 
                    a.name as name_areas, 
                    count(e.id) as count_employers, 
                    count(v.id) as count_vacancies
                from areas a 
                join employers e on a.id = e.area
                left join vacancies v on a.id = v.area_id
                where v.id is null
                group by 1
                order by 3, 2 DESC
                limit 5
            '''

In [440]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name_areas  count_employers  count_vacancies
0              Россия              410                0
1           Казахстан              207                0
2  Московская область               75                0
3  Краснодарский край               19                0
4            Беларусь               18                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [441]:
query_5_3 = f'''select 
                    e.name as name_employer, 
                    count(distinct v.area_id) as count_areas
                from vacancies v
                join employers e on v.employer_id = e.id
                group by 1 
                order by 2 DESC
                limit 5
            '''

In [442]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name_employer  count_areas
0                  Яндекс          181
1              Ростелеком          152
2              Спецремонт          116
3  Поляков Денис Иванович           88
4               ООО ЕФИН            71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [443]:
query_5_4 = f'''select count(e.name) as count_employers
                from employers e
                left join employers_industries ei on e.id = ei.employer_id 
                where ei.industry_id IS NULL
            '''

In [444]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count_employers
0             8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [445]:
query_5_5 = f'''select e.name as name_employers
                from employers e
                join employers_industries ei on e.id = ei.employer_id
                group by 1
                having count(ei.industry_id) = 4
                order by 1 asc 
                limit 5
            '''

In [446]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name_employers
0   101 Интернет
1       21vek.by
2           2ГИС
3             2К
4    4 пикселя +

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [447]:
query_5_6 = f'''select count(*) as count_employers
                from employers_industries ei
                join industries i on ei.industry_id = i.id
                where i.name = 'Разработка программного обеспечения'
            '''

In [448]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count_employers
0             3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [449]:
import pandas as pd
import requests

url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
table = pd.read_html(url)
mill_cities = pd.DataFrame(table[1])
mill_cities = tuple(mill_cities['Город'].replace('Воронеж[a]', 'Воронеж'))
mill_cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [467]:
query_5_7= f'''(select a.name as name, count(*) cnt
                from employers e
                join vacancies v on e.id = v.employer_id
                join areas a on a.id = v.area_id
                where e.name = 'Яндекс' and a.name in {mill_cities}
                group by a.id
                order by 2 asc)
            
                UNION ALL 

                (select 'TOTAL', count (*) 
                from vacancies v
                join employers e ON e.id = v.employer_id
                join areas a ON a.id = v.area_id
                where e.name = 'Яндекс' and a.name in {mill_cities})            
           '''

In [468]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0             TOTAL  485
1              Омск   21
2         Челябинск   22
3        Красноярск   23
4         Волгоград   24
5    Ростов-на-Дону   25
6            Казань   25
7             Пермь   25
8            Самара   26
9               Уфа   26
10        Краснодар   30
11          Воронеж   32
12      Новосибирск   35
13  Нижний Новгород   36
14     Екатеринбург   39
15  Санкт-Петербург   42
16           Москва   54

***

- На первом месте по количеству вакансий находится компания "Яндекс", кроме того, данная компания является лидером по количеству регионов, в которых она публикует свои вакансий. Предположительно это обусловлено наличием большой филиальной структуры и вакансиями, у которых указан тип рабочего графика "Удаленная работа". Вакансии компании "Яндекс" предстваленны в 16 городах-миллионниках, самое большое количество вакансий предствалено в г. Моска и Санкт-Петербург, меньше всего вакансий в г.Омск. 
- Наибольшее количество работодателей, у которых отсутствуют вакансии находится в России. 
35% работодателей (8419) не указывают сферу деятельности, однако есть работодатели, которые указывают 4 сферы деляьности. 
- Сферу деятельности "Разработка программного обеспечения" указывают 3553 или 15 % работодателей.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [469]:
query_6_1 = f'''select count(*) as count_v_ds
                from vacancies
                where lower(name) like '%data%' OR lower(name) like '%данн%'
            '''

In [470]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count_v_ds
0        1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [471]:
query_6_2 = f'''select count(name) as cnt_v_ds
                from vacancies
                where (name ilike '%data%scientist%' 
                or name ilike '%data%science%'
                or name ilike '%исследователь%данных%'
                or (name like '%ML%' and name not ilike '%HTML%')
                or name ilike '%machine%learning%'
                or name ilike '%машинн%обучен%%')
                and (name ilike '%junior%'
                or experience = 'Нет опыта'
                or employment = 'Стажировка')
            '''

In [472]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


cnt_v_ds
0        51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [473]:
query_6_3 = f'''select count(name) as cnt_ds
                from vacancies
                where (name ilike '%data%scientist%' 
                or name ilike '%data%science%'
                or name ilike '%исследователь%данных%'
                or (name like '%ML%' and name not ilike '%HTML%')
                or name ilike '%machine%learning%'
                or name ilike '%машинн%обучен%%')
                and (key_skills ilike '%SQL%'
                or key_skills ilike '%postgres%')
            '''

In [474]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


cnt_ds
0     201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [475]:
query_6_4 = f'''select count(name) as cnt_ds_py
                from vacancies
                where (name ilike '%data%scientist%' 
                or name ilike '%data%science%'
                or name ilike '%исследователь%данных%'
                or (name like '%ML%' and name not ilike '%HTML%')
                or name ilike '%machine%learning%'
                or name ilike '%машинн%обучен%%')
                and (key_skills ilike '%Python%')
            '''

In [476]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


cnt_ds_py
0        351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [477]:
query_6_5 = f''' select round(avg(1 + length(key_skills) - length(replace(key_skills, CHR(9), ''))), 2) as key_skills
                 from vacancies
                 where (name ilike '%data%scientist%' 
                 or name ilike '%data%science%'
                 or name ilike '%исследователь%данных%'
                 or (name like '%ML%' and name not ilike '%HTML%')
                 or name ilike '%machine%learning%'
                 or name ilike '%машинн%обучен%%')
                 and  (key_skills is not null)
              '''

In [478]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\390248827.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


key_skills
0        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [479]:
query_6_6 = f'''select experience,
                        round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from)) / 2)) as avg_salary_ds
                from vacancies
                where (salary_to is not null
                or salary_from is not null) 
                and (name ilike '%data%scientist%' 
                or name ilike '%data%science%'
                or name ilike '%исследователь%данных%'
                or (name like '%ML%' and name not ilike '%HTML%')
                or name ilike '%machine%learning%'
                or name ilike '%машинн%обучен%%')
                group by experience
            '''


In [480]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  avg_salary_ds
0           Нет опыта        74643.0
1  От 1 года до 3 лет       139675.0
2       От 3 до 6 лет       243115.0

***

- Для начинающих DS предлагается не большое количество вакансий, а именно 51.
- Увеличение количесва вакансий зависит от ключевых навыков, указанных в вакансиях. Так при указании ключевых навыков "SQL" или "postgres" количество вакансий увеличивается до 201, а при указании ключевого навыка "Python" - до 351. 
- В среднем, в вакансиях для DS, указывается более 6 ключевых навыков. 
- Заработная плата для DS увеличивается более чем на 30% в зависимости от опыта работы. 


# Общий вывод по проекту

Каких вакансий много? 

In [486]:
query_7_1 = f'''select name as vacancy_name, count(*) as vacancy_count
                from vacancies
                group by 1
                order by 2 DESC
                limit 30
            '''

In [487]:
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_17484\2911634181.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


vacancy_name  vacancy_count
0                             Системный администратор           1761
1                                      Программист 1С           1153
2                                     Бизнес-аналитик            488
3                    Специалист технической поддержки            457
4                                  Системный аналитик            432
5                                            Аналитик            404
6                                 Инженер-программист            389
7                                      Программист 1C            360
8                                Frontend-разработчик            242
9                       Инженер технической поддержки            230
10                                        Программист            224
11                               Руководитель проекта            191
12            Младший специалист ручного тестирования            169
13        Специалист службы поддержки в Яндекс Бизнес            168
14                                  Менеджер проектов            153
15                                   Java-разработчик            150
16          Специалист по информационной безопасности            148
17  Специалист службы поддержки Рекламной сети Янд...            139
18                                    PHP-разработчик            135
19          Специалист службы поддержки пользователей            126
20                                Компьютерный мастер            123
21                                   Менеджер проекта            120
22                               Технический писатель            111
23                                Маркетолог-аналитик            108
24                              Руководитель проектов            106
25                             Ведущий программист 1C            106
26                             Технический специалист            104
27                 Сотрудник поддержки поиска Яндекса             99
28                    Ведущий системный администратор             99
29         Технический писатель (дистанционный автор)             99

Из содержащиъся в данных вакансиях, самое большое количество принадлежит вакансиям: Системный администратор и Программист 1С. 